dali/python/nvidia/dali/ops.py

```diff
@@ -613,19 +545,112 @@ def _load_ops():
-    _mixed_ops = _mixed_ops.union(set(b.RegisteredMixedOps()))
+    _cpu_ops = _cpu_ops.union(set(_b.RegisteredCPUOps()))
+    _gpu_ops = _gpu_ops.union(set(_b.RegisteredGPUOps()))
+    _cpu_ops.remove("ExternalSource")
+    _gpu_ops.remove("ExternalSource")
+    _mixed_ops = _mixed_ops.union(set(_b.RegisteredMixedOps()))
     _cpu_gpu_ops = _cpu_ops.union(_gpu_ops).union(_mixed_ops)
     for op_name in _cpu_gpu_ops:
         if not hasattr(sys.modules[__name__], op_name):
-            setattr(sys.modules[__name__], op_name,
-                    python_op_factory(op_name, op_device = "cpu"))
+            op_class = python_op_factory(op_name, op_device = "cpu")
+            setattr(sys.modules[__name__], op_name, op_class)
+            if op_name not in ["ExternalSource"]:
+                _functional._wrap_op(op_class)
+
 _load_ops()
 
 def Reload():
     _load_ops()
 
+
+# custom wrappers around ops
+class ExternalSource(with_metaclass(_DaliOperatorMeta, object)):
+    global _cpu_ops
+    global _gpu_ops
+    _cpu_ops = _cpu_ops.union({'ExternalSource'})
+    _gpu_ops = _gpu_ops.union({'ExternalSource'})
+
+    def __init__(self, callback = None, num_outputs = None, *, layout = None, name = None, device = "cpu", **kwargs):
+        self._schema = _b.GetSchema("_ExternalSource")
+        self._spec = _b.OpSpec("_ExternalSource")
+        self._device = device
+        self._layout = layout
+
+        if name is not None and self.num_outputs is not None:
+            raise ValueError("`num_outputs` is not compatible with named `ExternalSource`")
+
+        self._name = name
+        self._num_outputs = num_outputs
+        self._callback = callback
+
+        self._spec.AddArg("device", device)
+        for key, value in kwargs.items():
+            self._spec.AddArg(key, value)
+
+    @property
+    def spec(self):
+        return self._spec
+
+    @property
+    def schema(self):
+        return self._schema
+
+    @property
+    def device(self):
+        return self._device
+
+    @property
+    def preserve(self):
+        return False
+
+    def __call__(self, *, callback = None, name = None, **kwargs):
+        if callback is None:
+            callback = self._callback
+        else:
+            if self._callback is not None:
+                raise RuntimeError("The callback for external source was specified twice. "
+                    "Callback must not be passed both to constructor and call.")
+
+        if name is None:
+            name = self._name
+
+        if name is not None and self._num_outputs is not None:
+            raise RuntimeError("`num_outputs` is not compatible with named `ExternalSource`")
+
+
+        if self._num_outputs is not None:
+            outputs = []
+            kwargs = {}
+            group = []
+            for i in range(self._num_outputs):
+                op_instance = _OperatorInstance([], self, **kwargs)
+                op_instance._callback = callback
```

Comment:
I don't get what is happening with the callback if we don't pass any and it's set to None.